code/hsec-sync/src/Security/Advisories/Sync/Snapshot.hs

```diff
@@ -0,0 +1,181 @@
+            DirectorySetupSnapshotArchive x -> "Directory setup got an exception: " <> show x
+            ExtractSnapshotArchive x -> "Extraction got an exception: " <> show x
+        ]
+
+data Snapshot = Snapshot
+  { snapshotRoot :: FilePath,
+    repositoryUrl :: String,
+    repositoryBranch :: String
+  }
+
+snapshotArchiveUrl :: Snapshot -> String
+snapshotArchiveUrl s =
+  ensureFile (mkUrl [repositoryUrl s, "archive/refs/heads", repositoryBranch s]) <> ".tar.gz"
+
+data SnapshotRepositoryStatus
+  = SnapshotDirectoryMissing
+  | SnapshotDirectoryInitialized
+  | SnapshotDirectoryIncoherent
+
+snapshotRepositoryStatus :: Snapshot -> IO SnapshotRepositoryStatus
+snapshotRepositoryStatus s = do
+  dirExists <- D.doesDirectoryExist $ snapshotRoot s
+  if dirExists
+    then do
+      dirAdvisoriesExists <- D.doesDirectoryExist $ snapshotRoot s </> "advisories"
+      fileMetadataExists <- D.doesFileExist $ snapshotRoot s </> "snapshot.json"
+      return $
+        if dirAdvisoriesExists && fileMetadataExists
+          then SnapshotDirectoryInitialized
+          else SnapshotDirectoryIncoherent
+    else return SnapshotDirectoryMissing
+
+data SnapshotRepositoryEnsuredStatus
+  = SnapshotRepositoryCreated
+  | SnapshotRepositoryExisting
+
+ensureSnapshot ::
+  Snapshot ->
+  SnapshotRepositoryStatus ->
+  IO (Either SnapshotError SnapshotRepositoryEnsuredStatus)
+ensureSnapshot s =
+  \case
+    SnapshotDirectoryMissing ->
+      ($> SnapshotRepositoryCreated)
+        <$> overwriteSnapshot s
+    SnapshotDirectoryIncoherent ->
+      ($> SnapshotRepositoryCreated)
+        <$> overwriteSnapshot s
+    SnapshotDirectoryInitialized ->
+      return $ Right SnapshotRepositoryExisting
+
+overwriteSnapshot :: Snapshot -> IO (Either SnapshotError ())
+overwriteSnapshot s =
+  runExceptT $ do
+    let root = snapshotRoot s
+    ensuringPerformed <- liftIO $ try $ ensureEmptyRoot root
+    whenLeft ensuringPerformed $
+      throwE . SnapshotProcessError . DirectorySetupSnapshotArchive
+
+    resultE <- liftIO $ try $ get $ snapshotArchiveUrl s
+    case resultE of
+      Left e ->
+        throwE $
+          SnapshotProcessError $
+            FetchSnapshotArchive $
+              case e of
+                InvalidUrlException url reason ->
+                  "Invalid URL " <> show url <> ": " <> show reason
+                HttpExceptionRequest _ content ->
+                  case content of
+                    StatusCodeException response body ->
+                      "Request failed with " <> show (response ^. responseStatus) <> ": " <> show body
+                    _ ->
+                      "Request failed: " <> show content
+      Right result -> do
+        performed <-
+          liftIO $
+            try $
+              withSystemTempDirectory "security-advisories" $ \tempDir -> do
+                let archivePath = tempDir <> "/snapshot-export.tar.gz"
+                BL.writeFile archivePath $ result ^. responseBody
```

Comment:
why not just `responseBody result`? https://hackage.haskell.org/package/http-client-0.7.17/docs/Network-HTTP-Client.html#v:responseBody

Reply:
that's how `wreq` suggest to do it.

Reply:
yuh, I realised, I guess if you wanna use that instead of `https://hackage.haskell.org/package/http-conduit-2.3.8.3/docs/Network-HTTP-Simple.html` like in my PR, then this is the only way; I don't know what it does justifies the `lens` dependency, too.

Reply:
but it's a matter of taste